### PR TITLE
Fix ImageContainer2D for Fury v2: wrap image in UI and handle RGB images

### LINKED ISSUE #1108: Port `ImageContainer2D` Component to v2
**Is your feature request related to a problem?**
We are porting existing UI components to the v2 branch to remains consistent and functional as we transition to the new architecture.

**Describe the solution you'd like**
Port `ImageContainer2D` to the v2 branch. The implementation should follow the new architecture present in v2 branch while almost preserving the original API and user experience.

Key Requirements:
- Implementation: Port the core logic from the legacy branch into the v2.
- Dependencies: Identify and port any required sub-components or internal helper classes necessary for this UI to function correctly.
- Testing: Port and adapt all existing and create new tests associated with this component to verify features.

**Additional context**
- Base Branch: v2
- Reference the existing v2 UI components for the standards and architecture. eg. Panel2D

### PR DESCRIPTION
**Pull request by VanshAgarwal24036**
This PR fixes and ports `ImageContainer2D` to work properly with the current Fury v2 architecture.

Fixes #1108 

This PR:

- Wraps the image actor inside a `_ImageUI` class that properly subclasses `UI`.
- Implements the required `_update_actors_position` method.
- Converts RGB images to grayscale to ensure compatibility with the underlying image actor.
- Ensures correct actor positioning using the new UI system.
- Adds a dedicated test for `ImageContainer2D`.

### CODE REVIEW

**Comment by VanshAgarwal24036** [comment]
Hi maintainers 

All tests and ruff checks pass locally.
Kindly approve the workflows so CI can run.

Thank you!

**Comment by VanshAgarwal24036** [comment]
@maharshi-gor  @ganimtron-10 plz review

**Comment by m-agour** [comment]
Hi @VanshAgarwal24036 , thanks for this PR, but, we are currently develpoing the v2 branch not the master branch, and given there is no conflicts with master I don't think is is based on v2, please always start your branch based on v2 branch not master.

**Comment by VanshAgarwal24036** [comment]
Thankyou @m-agour I will review again and do changes accordingly to v2 branch

**Comment by VanshAgarwal24036** [comment]
@m-agour please check now

File: fury/ui/containers.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 import numpy as np
 
+from fury import actor as fury_actor, io
```
Comment (maharshi-gor) [inline fury/ui/containers.py]: Not the correct way to import fury actors

Please use from fury.actor import image as image_actor if you need to.

File: fury/ui/containers.py [hunk 2]
```diff
@@ -979,159 +980,40 @@ def border_width(self, side_width):
...
-# #         self.actor.SetPosition(*coords)
-
-# #     def scale(self, factor):
-# #         """Scale the image.
-
-# #         Parameters
-# #         ----------
-# #         factor : (float, float)
-# #             Scaling factor (width, height) in pixels.
+    def __init__(self, image_input, position=(0, 0)):
```
Comment (maharshi-gor) [inline fury/ui/containers.py]: support for size.

**Review by maharshi-gor — CHANGES_REQUESTED** [review]
Thank you @VanshAgarwal24036 for the work.

The PR requires better doc-strings and few feature support.

**Review by ganimtron-10 — CHANGES_REQUESTED** [review]
Hi @VanshAgarwal24036,

Thanks for the PR. I noticed you closed a previous PR for the same thing, just fyi you don't need to create a new PR for new changes. If you commit to the same branch, it will automatically reflect in the existing PR.

I haven't done a deep dive yet, but here are my initial thoughts:
1. Please include a demo using ImageContainer2D to make it easier to test the functionality.
2. Don't create a separate file for tests, they should follow the existing project structure.
3. Try to add more tests to validate all aspects of the UI, not just the happy path.

Please update the PR accordingly. Thanks!

**Comment by VanshAgarwal24036** [comment]
Thanks for the helpful feedback!

I’ve updated the PR according to the suggestions:

* Fixed the actor import as recommended.
* Added support for the `size` parameter.
* Added a demo example for `ImageContainer2D`.
* Moved the tests to follow the existing project test structure and added more cases.

Also, since this is my first contribution to FURY, I initially got confused when the PR showed a very large number of commits. After looking into it, I realized that I had mistakenly based my branch on the wrong branch instead of `v2`. I closed the earlier PR and recreated the changes correctly based on the `v2` branch.

Please let me know if any further improvements are needed. Thanks again for the review!

**Comment by VanshAgarwal24036** [comment]
@ganimtron-10 @maharshi-gor please tell me how to start workflow waiting

**Comment by VanshAgarwal24036** [comment]
@maharshi-gor As my PR is failing with 20 checks so on your suggestion to @ganimtron-10 on his PR to rebase the branch and the checks pass as same problem in my PR also so now i have rebased so Please review it.

**Comment by VanshAgarwal24036** [comment]
Hi @maharshi-gor and @ganimtron-10 

I have rebased the branch on the latest `v2` as suggested and updated the PR.
All CI checks should now run against the updated base.

Could you please take another look when you get time?
Thanks for your guidance!

File: fury/ui/containers.py [hunk 3]
```diff
@@ -979,159 +981,56 @@ def border_width(self, side_width):
...
-# #         """Add all subcomponents or VTK props that compose this UI component.
-
-# #         Parameters
-# #         ----------
-# #         scene : scene
-
-# #         """
-# #         scene.add(self.actor)
-
-# #     def resize(self, size):
```
Comment (ganimtron-10) [inline fury/ui/containers.py:1088]: resize method is missing

File: fury/ui/containers.py [hunk 4]
```diff
@@ -979,159 +981,56 @@ def border_width(self, side_width):
...
+            )
+        if image.ndim == 3:
+            image = image.mean(axis=2).astype(np.uint8)
+        if image.ndim != 2:
+            raise ValueError("Only 2D images are supported.")
+        if size is not None and len(size) != 2:
+            raise ValueError("size must be (width, height)")
+        self._img = image
+        self._size = size
+        self._image_actor = image_actor(self._img)
```
Comment (ganimtron-10) [inline fury/ui/containers.py]: this should be done in _setup()

File: fury/ui/containers.py [hunk 5]
```diff
@@ -979,159 +981,56 @@ def border_width(self, side_width):
...
+        super().__init__(position=position)
 
-# #         """
-# #         self.resize(self.size * factor)
+    def _setup(self):
+        self._actors = [self._image_actor]
```
Comment (ganimtron-10) [inline fury/ui/containers.py:1021]: why is this required?

File: fury/ui/tests/test_containers.py [hunk 6]
```diff
@@ -179,6 +180,37 @@ def test_panel2d_visual_snapshot(
...
+    assert container is not None
+    assert len(container.actors) == 1
+
+
+def test_image_container_rgb():
+    """Test that RGB images are converted to grayscale."""
+    img = np.zeros((10, 10, 3), dtype=np.uint8)
+    container = ui.ImageContainer2D(img)
+
+    assert container is not None
```
Comment (ganimtron-10) [inline fury/ui/tests/test_containers.py]: where are we testing rgb converted to grayscale?

File: fury/ui/tests/test_containers.py [hunk 7]
```diff
@@ -179,6 +180,37 @@ def test_panel2d_visual_snapshot(
...
+
+    assert container._get_size() == (50, 50)
+
+
+def test_image_container_invalid_input():
+    """Test invalid input type raises TypeError."""
+    with pytest.raises(TypeError):
+        ui.ImageContainer2D(123)
+
+
```
Comment (ganimtron-10) [inline fury/ui/tests/test_containers.py:222]: tests can be improved and added to test functionality and edge cases

**Review by ganimtron-10 — CHANGES_REQUESTED** [review]
Hi @VanshAgarwal24036 ,
Thanks for the PR!

PTAL at the below comments.

File: fury/ui/containers.py (hunk 5 above)
Comment (VanshAgarwal24036) [reply fury/ui/containers.py:1021]: As `ImageContainer2D` wrapsVTK image actor, `_setup` assign it to `self._actors` so base UI class can manage it properly for rendering, event handling and positioning.

**Comment by VanshAgarwal24036** [comment]
Updated the implementation to align with the current backend constraints.
RGB inputs are now converted to grayscale internally since the image actor supports only 2D images.
Simplified the pipeline and updated tests + example accordingly.

File: docs/examples/viz_image_container.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+##############################################################################
+# Creating sample images
+
+# Grayscale image (used as-is)
+img_gray = np.random.randint(0, 255, (200, 200), dtype=np.uint8)
+
+# RGB image (will be converted internally to grayscale)
+img_rgb = np.random.randint(0, 255, (200, 200, 3), dtype=np.uint8)
+
+##############################################################################
```
Comment (maharshi-gor) [inline docs/examples/viz_image_container.py:48]: I would use one of the texture images provided by fury.data.

you can do read viz_skybox example to check.

**Review by maharshi-gor — CHANGES_REQUESTED** [review]
Hey @VanshAgarwal24036 Thank you for the work.

Please check my comments. I see good progress on the PR.

**Comment by VanshAgarwal24036** [comment]
@maharshi-gor I have update the docs I think due to which Ci fail

**Comment by ganimtron-10** [comment]
Hi @VanshAgarwal24036,
I tried executing the demo and its not working.

Please see the below error:
image

**Comment by VanshAgarwal24036** [comment]
Hi @ganimtron-10 

Thanks for the review. I’ve addressed the issues and updated the implementation accordingly.

Key updates:

* Fixed image handling to support RGB/RGBA properly
* Cleaned up grayscale conversion logic
* Ensured demo works correctly with provided assets

Please take another look and let me know if anything else needs improvement.

File: docs/examples/viz_image_container.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+# Fetch icons
+
+fetch_viz_icons()
+
+##############################################################################
+# Load image properly
+
+img_path = read_viz_icons(fname="circle-right.png")
```
Comment (maharshi-gor) [inline docs/examples/viz_image_container.py]: Please load something better

**Review by maharshi-gor — CHANGES_REQUESTED** [review]
Hey @VanshAgarwal24036 thanks for updating the image is not very useful understanding the actor is good or not in the tutorial please pick any texture files. which would be better.

**Comment by VanshAgarwal24036** [comment]
@maharshi-gor I’ve update the example to use a better texture with clearer details and a proper aspect ratio for improved visualization.

Previously, some textures led to distorted or unclear rendering due to size/aspect ratio issues, so I’ve pick a more suitable one for demonstration.

**Comment by ganimtron-10** [comment]
Hi @VanshAgarwal24036 ,

Is this the expected output?
image

If yes, could you help us understand, what should we/user deduce from this demo? How does it helps understand what ImageContainer2D is or what is its usecase?

**Comment by VanshAgarwal24036** [comment]
Hi @ganimtron-10

No, this is not the expected final output.

I’m currently using textures from the dataset, but I’m facing two issues:

- Many textures are very large or panoramic, which leads to stretched or unclear output (as shown in the screenshot)
- Some textures also trigger:

`D:\gsoc\fury\fury\venv311\Lib\site-packages\PIL\Image.py:3451: DecompressionBombWarning: Image size (131220000 pixels) exceeds limit of 89478485 pixels, could be decompression bomb DOS attack.
  warnings.warn(`

image

Because of this, the example is not clearly demonstrating the intended behavior of ImageContainer2D.

Intended purpose of this demo:

- Show rendering of a 2D image in screen (UI) space
- Demonstrate positioning and sizing independent of the 3D scene
- Illustrate UI use cases like overlays or HUD elements

Could you suggest:

- a preferred texture from the dataset, or
- the expected type of image to use for this example?

I’ll proceed based on your guidance.

**Comment by maharshi-gor** [comment]
Hey @VanshAgarwal24036  you can get the skybox texture it is made of 6 images just use one of them.

**Comment by maharshi-gor** [comment]
I am closing this as I do not see any answer on this.
